### DISCUSSION TITLE
What ObjectId to use?

### DISCUSSION THREAD

**Discussion by wasya243** [typegoose/typegoose#841] [Q&A]
Hello, I just wondering what ObjectId to use when defining typegoose class.

```ts
import * as mongoose from 'mongoose';
import { prop } from '@typegoose/typegoose';

class Foo {
  @prop()
  userId: mongoose.Types.ObjectId;
}
```

or

```ts
import * as mongoose from 'mongoose';
import { prop } from '@typegoose/typegoose';

class Foo {
  @prop()
  userId: mongoose.Schema.Types.ObjectId;
}
```

Yes, I've already seen this https://typegoose.github.io/typegoose/docs/guides/advanced/using-objectid-type. But just wondering why. In mongoose docs you should use mongoose.Schema.Types.ObjectId in schemas and here you should use a different thing.

**Comment by hasezoey** [comment]
in typegoose always use `mongoose.Types.ObjectId`, the reason being that the type defined on the class is also the type that will be used in the document type (see example), also typegoose / mongoose (i dont remember which, but it is done) automatically translates `mongoose.Types.*` to `mongoose.Schema.Types.*` at runtime 

(the example is not using `@prop` because decorators do not affect the type and this is a example about types)
```ts
class Test {
  public t1: mongoose.Schema.Types.ObjectId;
  public t2: mongoose.Types.ObjectId;
}

type doc = DocumentType<Test>;
doc.t1; // type is "Schema.Types.ObjectId", which is not useful or correct
doc.t2; // type is "Types.ObjectId" which is the correct type
```

mongoose works around this by having the `InferSchemaType` translate the types or by having the manual interfaces also already as `mongoose.Schema.Types`
